### PR TITLE
0211-maintain

### PR DESCRIPTION
**Pull request by cyw0ng95**
- Add provider-type compatibility validation (NVD→CVE, MITRE→CWE/CAPEC/ATT&CK, SSG→SSG)
- Add atomic ID format validation using regex patterns (CVE-YYYY-NNNN, CWE-N, CAPEC-N, TNNNN)
- Add atomic ID maximum length validation (256 chars)
- Add edge case validation (empty strings, whitespace, empty parts between separators)
- Update existing test to use valid CVE ID format
- Add comprehensive test coverage for new validation rules

### CODE REVIEW

**Comment by MonkeyCode-AI** [comment]
MonkeyCode-AI 正在分析任务...

File: pkg/urn/urn.go [hunk 1]
```diff
@@ -64,11 +82,24 @@ type URN struct {
...
 	if len(parts) != 4 || parts[0] != "v2e" {
 		return nil, fmt.Errorf("%w: expected format 'v2e::<provider>::<type>::<atomic_id>', got '%s'", ErrInvalidURN, s)
 	}
 
+	// Check for empty parts between :: (e.g., "v2e::nvd::::test")
+	for i, part := range parts {
+		if part == "" {
+			return nil, fmt.Errorf("%w: empty part at position %d", ErrInvalidURN, i)
+		}
+	}
```
Comment (cyw0ng95) [inline pkg/urn/urn.go:101]: > [!WARNING]
> **⚠️** Parse 对空 atomicID 的错误语义改变：现在被“空 part”提前拦截为 ErrInvalidURN，无法再返回 ErrEmptyAtomicID
> 
> 当前 Parse 先 Split 再检查空 part（parts[i]==""）。对于输入 `v2e::nvd::cve::`，parts[3] 为空，会在“empty part”检查处直接返回 ErrInvalidURN，而后面的 `if atomicID == "" { return ErrEmptyAtomicID }` 永远不会触发。这会造成 Parse 与 New 的错误语义不一致：New 对空 atomicID 仍返回 ErrEmptyAtomicID，但 Parse 返回 ErrInvalidURN。并且这让 ErrEmptyAtomicID 在 Parse 路径中失去意义（除非未来移除该空 part 检查或变更分割策略）。
> 
> **建议**: 建议在 Parse 中仅对 provider/type 段做“空 part”检查，允许 atomicID 段为空从而走到专门的 ErrEmptyAtomicID 分支；或反过来统一约定：atomicID 为空属于 URN 结构错误（ErrInvalidURN），并相应调整 New 也返回 ErrInvalidURN/ErrEmptyAtomicID 的一致策略。
> 
```suggestion
func Parse(s string) (*URN, error) {
	// Trim whitespace
	s = strings.TrimSpace(s)
	if s == "" {
		return nil, fmt.Errorf("%w: empty URN string", ErrInvalidURN)
	}

	parts := strings.Split(s, "::")
	if len(parts) != 4 || parts[0] != "v2e" {
		return nil, fmt.Errorf("%w: expected format 'v2e::<provider>::<type>::<atomic_id>', got '%s'", ErrInvalidURN, s)
	}

	// Check for empty parts between ::, but allow empty atomicID to be handled by ErrEmptyAtomicID
	if parts[1] == "" {
		return nil, fmt.Errorf("%w: empty part at position %d", ErrInvalidURN, 1)
	}
	if parts[2] == "" {
		return nil, fmt.Errorf("%w: empty part at position %d", ErrInvalidURN, 2)
	}

	provider := Provider(parts[1])
	resourceType := ResourceType(parts[2])
	atomicID := parts[3]

	// Validate provider
	if !isValidProvider(provider) {
		return nil, fmt.Errorf("%w: '%s' is not a valid provider", ErrInvalidProvider, provider)
	}

	// Validate resource type
	if !isValidType(resourceType) {
		return nil, fmt.Errorf("%w: '%s' is not a valid resource type", ErrInvalidType, resourceType)
	}

	// Validate provider-type compatibility
	if !isValidProviderTypeCombination(provider, resourceType) {
		return nil, fmt.Errorf("%w: provider '%s' cannot provide resource type '%s'", ErrProviderTypeMismatch, provider, resourceType)
	}

	// Validate atomic ID
	if atomicID == "" {
		return nil, ErrEmptyAtomicID
	}

	// Check atomic ID length
	if len(atomicID) > maxAtomicIDLength {
		return nil, fmt.Errorf("%w: length %d exceeds maximum %d", ErrAtomicIDTooLong, len(atomicID), maxAtomicIDLength)
	}

	// Validate atomic ID format based on type
	if err := validateAtomicIDFormat(resourceType, atomicID); err != nil {
		return nil, err
	}

	return &URN{
		Provider: provider,
		Type:     resourceType,
		AtomicID: atomicID,
	}, nil
}
```

File: pkg/urn/urn.go [hunk 2]
```diff
@@ -64,11 +82,24 @@ type URN struct {
 //
 // Expected format: v2e::<provider>::<type>::<atomic_id>
 func Parse(s string) (*URN, error) {
+	// Trim whitespace
+	s = strings.TrimSpace(s)
+	if s == "" {
```
Comment (cyw0ng95) [inline pkg/urn/urn.go:87]: > [!WARNING]
> **⚠️** Parse 仅 trim 整个 URN 字符串，但不 trim atomicID；New 会 trim atomicID，二者行为不一致
> 
> Parse 对输入整体 TrimSpace，但 atomicID 可能仍包含前后空格（例如 `v2e::ssg::ssg::  test  `），会被当作合法 atomicID（尤其 TypeSSG 不做格式限制）。而 New 会对 atomicID TrimSpace。这样同一个“逻辑 URN”通过 Parse 与 New 得到的 AtomicID 可能不同，容易导致存储/比较/去重不一致。
> 
> **建议**: 建议在 Parse 中也对 atomicID（至少）进行 TrimSpace，保持与 New 一致；如担心影响语义，则至少对非 SSG 类型强制 trim 并拒绝包含空格的 atomicID。
> 
```suggestion
atomicID := strings.TrimSpace(parts[3])
```

**Review by cyw0ng95 — COMMENTED** [review]
> 我是 [MonkeyCode AI 编程助手](https://monkeycode-ai.com/)，你可以在 GitHub 仓库的 PR 中 at @MonkeyCode-AI 来呼唤我。

> 任务执行细节请参考: https://monkeycode-ai.com/tasks/public?id=a63158b9-6bd0-44ea-a6ab-350fc436e1c1

## 代码审查结果

URN 解析与构造新增了较完整的校验并补齐测试，但存在 Parse/New 行为不一致及少量边界与测试设计问题需修复。


## ✨ 代码亮点

- 在 Parse 入口增加 TrimSpace 与空字符串处理，提升对用户输入的健壮性
- 补齐 provider/type 兼容性校验与 atomicID 格式、长度校验，能更早发现坏数据
- 新增较全面的测试覆盖（provider/type、格式、边界、长度、New trimming）

| 🚨 Critical | ⚠️ Warning | 💡 Suggestion |
|:-:|:-:|:-:|
| 0 | 2 | 0 |

**Comment by MonkeyCode-AI** [comment]
MonkeyCode-AI 正在分析任务...

**Review by cyw0ng95 — COMMENTED** [review]
> 我是 [MonkeyCode AI 编程助手](https://monkeycode-ai.com/)，你可以在 GitHub 仓库的 PR 中 at @MonkeyCode-AI 来呼唤我。

> 任务执行细节请参考: https://monkeycode-ai.com/tasks/public?id=332f5b74-db1e-4322-9709-5ba29bb5340b

## 代码审查结果

URN 校验与测试覆盖整体不错，但 ATT&CK 正则过宽、provider-type 强约束可能造成破坏性兼容问题，且单测错误断言方式不稳健，需修改后再合并。


## ✨ 代码亮点

- URN.Parse/New 增加 whitespace/空字符串/空分段/长度上限等边界校验，能显著提前拦截脏数据
- atomic ID 的类型化正则校验 + provider/type 兼容性校验，使 URN 约束更明确
- attack 模块为 TODO-007 增加了大量防 panic 测试与更健壮的 helper 实现，覆盖面较好

| 🚨 Critical | ⚠️ Warning | 💡 Suggestion |
|:-:|:-:|:-:|
| 0 | 1 | 0 |

File: pkg/urn/urn_test.go [hunk 3]
```diff
@@ -364,3 +364,296 @@ func TestRoundTrip(t *testing.T) {
...
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := Parse(tt.input)
+			if tt.wantErr != nil {
+				if err == nil {
+					t.Errorf("Parse() error = nil, wantErr %v", tt.wantErr)
+					return
+				}
+				if !strings.Contains(err.Error(), tt.wantErr.Error()) {
+					t.Errorf("Parse() error = %v, wantErr %v", err, tt.wantErr)
+				}
```
Comment (cyw0ng95) [inline pkg/urn/urn_test.go:426]: > [!WARNING]
> **⚠️** 单测通过 strings.Contains(err.Error(), wantErr.Error()) 判断错误类型不稳健
> 
> 多个新测试用 `strings.Contains(err.Error(), tt.wantErr.Error())` 来断言错误，这依赖错误字符串内容而非错误链，会在错误消息调整/本地化/包装层级变化时导致脆弱失败或误判。当前代码大量使用 `fmt.Errorf("%w: ...", ErrX)`，更合适的断言方式是 `errors.Is(err, tt.wantErr)`。
> 
> **建议**: 在测试文件中引入 `errors` 包，并将包含判断替换为 `errors.Is`。这样既能验证错误类型，也不会被错误文本变化影响。
>

**Comment by MonkeyCode-AI** [comment]
MonkeyCode-AI 正在分析任务...

File: pkg/rpc/client.go [hunk 4]
```diff
@@ -95,6 +107,13 @@ func (c *Client) HandleError(ctx context.Context, msg *subprocess.Message) (*sub
 
 // InvokeRPC invokes an RPC method on another service through the broker
 func (c *Client) InvokeRPC(ctx context.Context, target, method string, params interface{}) (*subprocess.Message, error) {
+	// Add panic recovery for broker disconnect scenarios
+	defer func() {
+		if r := recover(); r != nil {
+			c.logger.Error("Panic recovered in InvokeRPC: %v", r)
+		}
+	}()
+
```
Comment (cyw0ng95) [inline pkg/rpc/client.go:128]: > [!CAUTION]
> **🚨** InvokeRPC 的 panic recover 只记录日志但未将 panic 转为可返回错误，可能导致不可预期行为/掩盖并发问题
> 
> InvokeRPC 顶部新增 defer recover 仅记录日志，未通过命名返回值把 panic 转换为 error。panic 只有在 defer 中 recover 并让函数正常返回才能避免向上抛；当前写法可能导致执行路径中断且调用方语义不清晰。并且该 recover 可能掩盖 channel close 等并发问题的真实来源，增加排查难度。
> 
> **建议**: 要么移除该 recover（依赖显式错误分支处理断连/ctx done），要么把 InvokeRPC 改为命名返回值并在 recover 时设置 err 后返回，确保行为确定。
> 
```suggestion
func (c *Client) InvokeRPC(ctx context.Context, target, method string, params interface{}) (respMsg *subprocess.Message, err error) {
	// Convert panics into a normal error return
	defer func() {
		if r := recover(); r != nil {
			c.logger.Error("Panic recovered in InvokeRPC: %v", r)
			err = fmt.Errorf("panic in InvokeRPC: %v", r)
			respMsg = nil
		}
	}()

	// Generate correlation ID
	c.mu.Lock()
	c.correlationSeq++
	correlationID := fmt.Sprintf("%d", c.correlationSeq)
	entry := &RequestEntry{resp: make(chan *subprocess.Message, 1)}
	c.pendingRequests[correlationID] = entry
	c.mu.Unlock()

	// Clean up on exit
	defer func() {
		shouldClose := false
		c.mu.Lock()
		if _, ok := c.pendingRequests[correlationID]; ok {
			delete(c.pendingRequests, correlationID)
			shouldClose = true
		}
		c.mu.Unlock()
		if shouldClose {
			entry.Close()
		}
	}()

	payload, marshalErr := subprocess.MarshalFast(params)
	if marshalErr != nil {
		c.logger.Error("Failed to marshal RPC params: %v", marshalErr)
		return nil, marshalErr
	}

	msg := &subprocess.Message{
		Type:          subprocess.MessageTypeRequest,
		ID:            method,
		CorrelationID: correlationID,
		Source:        c.sp.Name(),
		Target:        target,
		Payload:       payload,
	}

	c.logger.Debug("Sending RPC request: method=%s, target=%s, correlationID=%s", method, target, correlationID)

	select {
	case <-c.sp.Context().Done():
		c.logger.Error("Subprocess context canceled before sending RPC request: method=%s, target=%s", method, target)
		return nil, fmt.Errorf("broker disconnected before RPC request could be sent")
	default:
	}

	if sendErr := c.sp.SendMessage(msg); sendErr != nil {
		c.logger.Error("Failed to send RPC request: %v", sendErr)
		return nil, sendErr
	}

	resp := entry.resp

	c.logger.Debug("Waiting for RPC response: method=%s, target=%s, correlationID=%s", method, target, correlationID)
	select {
	case response, ok := <-resp:
		if !ok {
			c.logger.Error("RPC response channel closed unexpectedly: method=%s, target=%s, correlationID=%s", method, target, correlationID)
			return nil, fmt.Errorf("broker disconnected during RPC call")
		}
		c.logger.Debug("Received RPC response: correlationID=%s, type=%s", correlationID, response.Type)
		return response, nil
	case <-time.After(c.rpcTimeout):
		timeoutErr := fmt.Errorf("RPC call timeout: method=%s, target=%s, correlationID=%s", method, target, correlationID)
		c.logger.Warn("RPC call timed out: %v", timeoutErr)
		return nil, timeoutErr
	case <-ctx.Done():
		ctxErr := ctx.Err()
		c.logger.Warn("RPC call context canceled while waiting for response: method=%s, target=%s, correlationID=%s, error: %v", method, target, correlationID, ctxErr)
		return nil, ctxErr
	case <-c.sp.Context().Done():
		c.logger.Error("Broker disconnected while waiting for RPC response: method=%s, target=%s, correlationID=%s", method, target, correlationID)
		return nil, fmt.Errorf("broker disconnected while waiting for RPC response")
	}
}
```

**Review by cyw0ng95 — COMMENTED** [review]
> 我是 [MonkeyCode AI 编程助手](https://monkeycode-ai.com/)，你可以在 GitHub 仓库的 PR 中 at @MonkeyCode-AI 来呼唤我。

> 任务执行细节请参考: https://monkeycode-ai.com/tasks/public?id=0fba05a7-b588-4c57-8899-b1c9cef37d0d

## 代码审查结果

整体实现了更严格的 URN 校验与 RPC 稳定性增强，但存在会导致崩溃/构建失败的关键问题（panic recover 语义与 defer 变量作用域），并有若干错误语义与测试稳健性需修正。


## ✨ 代码亮点

- URN: provider-type 兼容性校验 + atomic ID 正则校验 + 长度限制，覆盖面明显提升
- URN: Parse/New 对输入 trim、空字符串、空段等边界处理更严格
- RPC: RequestEntry.Signal/Close 增加并发 close 的 recover，配合 pendingRequests 清理意图上有助于避免 panic/泄漏
- ATT&CK / CVE / CWE / CAPEC 规则通过测试用例覆盖了常见有效/无效模式

| 🚨 Critical | ⚠️ Warning | 💡 Suggestion |
|:-:|:-:|:-:|
| 2 | 1 | 0 |

File: pkg/rpc/client.go [hunk 5]
```diff
@@ -110,12 +129,22 @@ func (c *Client) InvokeRPC(ctx context.Context, target, method string, params in
...
+		if exists {
+			delete(c.pendingRequests, correlationID)
+		}
 		c.mu.Unlock()
-		entry.Close()
+		// Only close if entry was still in map (not handled yet)
+		// If HandleResponse already processed it, it also called Signal which closes the channel
+		if exists {
+			entry.Close()
+		}
```
Comment (cyw0ng95) [inline pkg/rpc/client.go:159]: > [!CAUTION]
> **🚨** defer 清理逻辑中 exists 变量作用域错误，可能导致编译失败或清理逻辑失效
> 
> InvokeRPC 的 defer 闭包内用 `_, exists := ...` 声明 exists，但闭包外又使用 `if exists { entry.Close() }`。Go 作用域规则下 exists 仅在闭包内有效，按展示代码会出现 `undefined: exists` 编译错误；即便存在外部同名变量，也会导致预期不一致。该问题会阻塞构建或导致资源清理失效。
> 
> **建议**: 将 exists/ok 判断与 entry.Close() 放在同一个 defer 闭包内，或在 defer 外部提前声明并在闭包内赋值（注意并发）。推荐在闭包内计算 shouldClose 并在闭包内执行 Close。
> 
```suggestion
defer func() {
	shouldClose := false
	c.mu.Lock()
	if _, ok := c.pendingRequests[correlationID]; ok {
		delete(c.pendingRequests, correlationID)
		shouldClose = true
	}
	c.mu.Unlock()
	if shouldClose {
		entry.Close()
	}
}()
```

File: pkg/urn/urn_test.go (hunk 3 above)
Comment (cyw0ng95) [inline pkg/urn/urn_test.go:426]: > [!WARNING]
> **⚠️** 测试用 strings.Contains(err.Error(), wantErr.Error()) 判断错误类型不稳健
> 
> 使用 `strings.Contains(err.Error(), tt.wantErr.Error())` 断言错误类型对包装层级与 message 变更很敏感，可能误判/漏判。实现中已使用 `%w` 包装错误，更适合用 `errors.Is` 做类型断言；若需要再校验 message，可额外做 Contains。
> 
> **建议**: 将测试改为 `errors.Is(err, tt.wantErr)`；如需验证文案再加 message 断言。
> 
```suggestion
if tt.wantErr != nil {
	if err == nil {
		t.Fatalf("expected error %v, got nil", tt.wantErr)
	}
	if !errors.Is(err, tt.wantErr) {
		t.Fatalf("expected error %v, got %v", tt.wantErr, err)
	}
} else if err != nil {
	t.Fatalf("expected no error, got %v", err)
}
```

**Comment by MonkeyCode-AI** [comment]
MonkeyCode-AI 正在分析任务...

**Review by cyw0ng95 — COMMENTED** [review]
> 我是 [MonkeyCode AI 编程助手](https://monkeycode-ai.com/)，你可以在 GitHub 仓库的 PR 中 at @MonkeyCode-AI 来呼唤我。

> 任务执行细节请参考: https://monkeycode-ai.com/tasks/public?id=860fc36a-3e56-4c0d-9d70-156f3641abfb

## 代码审查结果

整体方向正确且测试覆盖提升明显，但 LearningFSM 解锁路径与 rpc.Client.InvokeRPC 的 panic 处理存在关键缺陷，需修复后再合并。


## ✨ 代码亮点

- URN Parse/New 增加 provider-type 兼容性校验与 atomicID 格式/长度限制，显著提升数据一致性
- URN 针对空字符串、空段、前后空白等边界情况的校验与测试覆盖较完整
- rpc.Client 在等待响应时额外监听 subprocess context Done，有助于更快感知 broker 断开并返回错误
- attack Excel 解析辅助函数增加 nil/越界保护，配套边界测试覆盖到位

| 🚨 Critical | ⚠️ Warning | 💡 Suggestion |
|:-:|:-:|:-:|
| 2 | 0 | 1 |

**Comment by MonkeyCode-AI** [comment]
MonkeyCode-AI 正在分析任务...

**Review by cyw0ng95 — COMMENTED** [review]
> 我是 [MonkeyCode AI 编程助手](https://monkeycode-ai.com/)，你可以在 GitHub 仓库的 PR 中 at @MonkeyCode-AI 来呼唤我。

> 任务执行细节请参考: https://monkeycode-ai.com/tasks/public?id=46e4217b-7b88-43ec-9a97-877a297f6b23

## 代码审查结果

本次变更提升了并发安全与健壮性，但 runenv.sh 存在容器清理逻辑缺陷与命令注入风险且 Optimizer 有潜在死锁点，需修复后再合入。


## ✨ 代码亮点

- URN.Parse/New 增加 provider-type 兼容性校验、atomicID 格式 regex 校验、长度上限与空/空白边界处理，且补充了覆盖度较高的测试用例
- LearningFSM 将 SaveState 从持锁状态中移出，并通过快照拷贝避免数据竞争，整体思路正确且更利于避免死锁
- Optimizer 将 batchSize 改为 atomic.Int32，并在 worker 读取 flushInterval 时做一致性读取，向并发安全迈进了一步
- rpc.Client 增强了 pendingRequests 清理与 broker disconnect 场景处理，并补充了针对 map 泄漏与并发 Close/Signal 的测试

| 🚨 Critical | ⚠️ Warning | 💡 Suggestion |
|:-:|:-:|:-:|
| 2 | 1 | 0 |

File: runenv.sh [hunk 6]
```diff
@@ -6,6 +6,45 @@
...
+if ! flock -n 200; then
+    echo "Waiting for another runenv.sh/build.sh instance to complete..."
+    if ! flock -w 1800 200; then
+        echo "Error: Timeout waiting for build lock after 30 minutes" >&2
+        echo "Another build may be stuck. Please check and manually remove $BUILD_LOCK_FILE if needed" >&2
+        exit 1
+    fi
+fi
+
+# Global variable to track container ID for cleanup
```
Comment (cyw0ng95) [inline runenv.sh:24]: > [!CAUTION]
> **🚨** CONTAINER_ID 从未赋值且使用 --rm，cleanup 中 stop/rm 逻辑无效且可能误导；同时 bash -c "$*" 存在注入风险
> 
> runenv.sh 新增 CONTAINER_ID 与 cleanup() 试图在退出时 podman stop/rm 容器，但脚本实际用的是 `podman run --rm ...`，且没有使用 `--name/--cidfile` 或捕获容器 ID，因此 CONTAINER_ID 永远为空，cleanup 的容器清理分支不会执行（即使执行也没意义，因为 --rm 会自动清理）。
> 
> 更严重的是：当传入命令参数时，脚本构造 `podman_base_cmd+=(bash -c "$*")`。这会把用户输入拼成一个 shell 字符串在容器内执行，若上层调用/CI 参数包含未转义内容，存在命令注入/意外解释风险（尤其当参数中包含 `;`, `&&`, `$()`, 反引号等）。
> 
> 此外，当前 trap 会在正常退出时调用 cleanup() 并 `exit $exit_code`，这会改变 run_container_env 的返回语义（函数内 `return $exit_code` 可能永远走不到），并且 cleanup 里强制 exit 可能导致上层逻辑难以维护。
> 
> **建议**: 二选一：
> 1) 如果继续使用 `podman run --rm`：删除 CONTAINER_ID 及 stop/rm 清理逻辑，仅保留 flock 解锁即可；并且避免 `bash -c "$*"`，改为以 argv 方式透传命令。
> 2) 如果需要显式 stop/rm：改为 `podman run -d --cidfile ...` 或 `--name` 捕获容器 ID，并在 cleanup 中基于 cidfile/name 清理，同时不要再用 --rm。
> 
> 同时，建议将容器命令执行改为：`podman ... -- "$@"`（不经 bash -c），或 `podman ... bash -lc` 但必须对参数进行严格转义。
> 
```suggestion
#!/bin/bash

# Run Environment script for v2e (Vulnerabilities Viewer Engine)
# This script creates the proper containerized build environment for macOS
# On Linux, it can optionally use a container with USE_CONTAINER=true

set -e

# Acquire exclusive lock to prevent parallel builds
# This protects shared resources (build artifacts, cache) from concurrent modification
BUILD_LOCK_FILE="${TMPDIR:-/tmp}/v2e-build.lock"
exec 200>"$BUILD_LOCK_FILE"

# Try to acquire lock with timeout (wait up to 30 minutes for another build to complete)
if ! flock -n 200; then
    echo "Waiting for another runenv.sh/build.sh instance to complete..."
    if ! flock -w 1800 200; then
        echo "Error: Timeout waiting for build lock after 30 minutes" >&2
        echo "Another build may be stuck. Please check and manually remove $BUILD_LOCK_FILE if needed" >&2
        exit 1
    fi
fi

# Release lock on exit (including error, interrupt, or termination)
cleanup() {
    flock -u 200 2>/dev/null || true
}

# Register cleanup function for all exit signals
trap cleanup EXIT INT TERM HUP QUIT

# Logging functions
log_info() {
    echo "-- $(date '+%H:%M:%S.%3N')/INFO/runenv -- $1"
}

log_error() {
    echo "-- $(date '+%H:%M:%S.%3N')/ERROR/runenv -- $1" >&2
    exit 1
}

# Global variables
BUILD_DIR=".build"
SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"

# Function to run container environment
run_container_env() {
    local os_name=$1
    shift  # Remove the first argument (os_name) to get remaining arguments
    log_info "Detected $os_name, starting container environment..."

    # Check if Podman is available
    if ! command -v podman &> /dev/null; then
        log_error "Podman is required but not installed or not in PATH"
        if [[ "$os_name" == "macOS" ]]; then
            log_info "Install Podman with: brew install podman"
        fi
        exit 1
    fi

    # Build container image with caching optimization
    log_info "Checking for existing development container image..."
    if ! podman images v2e-dev-container | grep -q "v2e-dev-container"; then
        log_info "Building development container from assets/dev.Containerfile..."
        if ! podman build -f assets/dev.Containerfile -t v2e-dev-container .; then
            log_error "Failed to build development container"
            exit 1
        fi
    else
        log_info "Using existing v2e-dev-container image"
    fi

    # Create Go module cache directory if it doesn't exist
    mkdir -p "${SCRIPT_DIR}/${BUILD_DIR}/pkg/mod"

    # Build podman command with common options
    podman_base_cmd=(podman run --rm \
        -v "$(pwd)":/workspace -w /workspace \
        -v "${SCRIPT_DIR}/${BUILD_DIR}/pkg/mod":/home/developer/go/pkg/mod \
        -e SESSION_DB_PATH="$SESSION_DB_PATH" \
        -e V2E_SKIP_WEBSITE_BUILD="$V2E_SKIP_WEBSITE_BUILD" \
        -e GO_TAGS="$GO_TAGS" \
        -e CGO_ENABLED="$CGO_ENABLED")

    # Add interactive flag if no command provided
    if [ $# -eq 0 ]; then
        podman_base_cmd+=(-it)
    fi

    # Add container image and command
    podman_base_cmd+=(v2e-dev-container)
    if [ $# -gt 0 ]; then
        # Pass argv through directly to avoid shell interpretation/injection
        podman_base_cmd+=("$@")
    fi

    "${podman_base_cmd[@]}"
    return $?
}

# Detect operating system
DETECTED_OS="$(uname -s)"

if [[ "$DETECTED_OS" == "Darwin" ]]; then
    # Always use container on macOS
    log_info "Detected macOS, running in containerized environment..."
    run_container_env "macOS" "$@"
elif [[ "$DETECTED_OS" == "Linux" ]]; then
    if [ "$USE_CONTAINER" = true ]; then
        # Use container on Linux if explicitly requested
        log_info "Container mode requested on Linux, running in containerized environment..."
        run_container_env "Linux" "$@"
    else
        # Run natively on Linux
        log_info "Detected Linux, running natively..."
        exec ./build.sh "$@"
    fi
else
    log_error "Unsupported platform: $DETECTED_OS. Only Linux and macOS are supported."
    exit 1
fi
```

File: cmd/v2broker/perf/optimizer.go [hunk 7]
```diff
@@ -190,6 +192,10 @@ func (o *Optimizer) EnableAdaptiveOptimization() {
 }
```
Comment (cyw0ng95) [inline cmd/v2broker/perf/optimizer.go:192]: > [!CAUTION]
> **🚨** EnableAdaptiveOptimization 回调中在持有 adaptationMu 时调用 applyAdaptedParameters，导致自锁死锁风险
> 
> EnableAdaptiveOptimization 的 callback 里：
> - 先 `o.adaptationMu.Lock()`
> - 满足频率后调用 `o.applyAdaptedParameters()`
> 而 applyAdaptedParameters 内部再次 `o.adaptationMu.Lock()`。
> 
> adaptationMu 是 sync.Mutex（不可重入），因此一旦进入该分支会发生死锁，导致监控回调卡死，进而影响自适应逻辑甚至阻塞相关 goroutine。
> 
> **建议**: 保证同一调用链中只加一次锁：
> - 要么在 callback 中不持锁调用 applyAdaptedParameters（只在 callback 中更新 lastAdaptation 等字段）；
> - 要么让 applyAdaptedParameters 假定锁已持有（改名 applyAdaptedParametersLocked，不再加锁）。
> 
> 同时，读取 flushInterval 时如果已经在持有同一把锁的范围内，可以直接读，避免额外锁开销。
>

File: pkg/rpc/client.go [hunk 8]
```diff
@@ -95,6 +116,16 @@ func (c *Client) HandleError(ctx context.Context, msg *subprocess.Message) (*sub
```
Comment (cyw0ng95) [inline pkg/rpc/client.go:116]: > [!WARNING]
> **⚠️** InvokeRPC 的 panicErr 永远不会生效：defer 设置的 panicErr 在函数返回后才执行
> 
> InvokeRPC 顶部声明 `var panicErr error` 并通过 `defer func(){... panicErr = ... }()` 试图捕获 panic。但该 defer 只会在函数返回时执行。
> 
> 当前函数内部 select 的各分支都会直接 `return`，如果发生 panic，defer 会恢复并设置 panicErr，但函数会继续沿着 panic 发生点之后的栈展开返回到调用方（实际上不会再执行到 select 之后的 `if panicErr != nil { ... }` 逻辑）。因此末尾的 panicErr 检查是不可达/无效的。
> 
> 此外，重复嵌套 recover（RequestEntry.Signal 里两层 recover）也显得多余，可简化。
> 
> **建议**: 将 panic recovery 的 defer 改为“具名返回值”模式，直接设置返回的 err：
> `func (c *Client) InvokeRPC(...) (respMsg *subprocess.Message, err error) { defer func(){ if r!=nil { err = fmt.Errorf(...) ; respMsg=nil } }() ... }`
> 这样才能确保 panic 被转换为 error 返回。
>

**Comment by MonkeyCode-AI** [comment]
MonkeyCode-AI 正在分析任务...

**Comment by cyw0ng95** [comment]
> 我是 [MonkeyCode AI 编程助手](https://monkeycode-ai.com/)，你可以在 GitHub 仓库的 PR 中 at @MonkeyCode-AI 来呼唤我。

> 任务执行细节请参考: https://monkeycode-ai.com/tasks/public?id=63b85761-ad44-4fce-ab99-2930f9b6ac61

⏳ **MonkeyCode-AI** 正在分析代码变更...

**Comment by MonkeyCode-AI** [comment]
MonkeyCode-AI 正在分析任务...

**Comment by cyw0ng95** [comment]
> 我是 [MonkeyCode AI 编程助手](https://monkeycode-ai.com/)，你可以在 GitHub 仓库的 PR 中 at @MonkeyCode-AI 来呼唤我。

> 任务执行细节请参考: https://monkeycode-ai.com/tasks/public?id=05678bfc-55ee-43ab-b72c-5b85fe3b5a21

⏳ **MonkeyCode-AI** 正在分析代码变更...

**Comment by MonkeyCode-AI** [comment]
MonkeyCode-AI 正在分析任务...

**Comment by cyw0ng95** [comment]
> 我是 [MonkeyCode AI 编程助手](https://monkeycode-ai.com/)，你可以在 GitHub 仓库的 PR 中 at @MonkeyCode-AI 来呼唤我。

> 任务执行细节请参考: https://monkeycode-ai.com/tasks/public?id=1a295fd8-ec64-4ce7-b33d-f4e3cf5118b9

⏳ **MonkeyCode-AI** 正在分析代码变更...
